### ISSUE TITLE
Full transaction pool drastically limits throughput

### ISSUE THREAD

**Issue by jakmeier** [near/nearcore#9326]
_edit:_ the root cause doesn't seem to be a full tx pool but some other (possibly related) reason why Sweat users hang forever

While testing with locust generated Sweat batches, I have observed full transaction pools and it results in vastly reduced throughput. Much more than I think is reasonable. It seems to make it really hard for new transactions to get in. And this is especially bad because clients don't even know that their transactions were dropped.

The image below shows how the TX pool fills up over ~ 2 days and stays full after that. (in a fork of testnet used for loadtesting onyl)

![image](https://github.com/near/nearcore/assets/6342444/82d6c374-1e7f-4eeb-be88-a2cdad5882bb)

And for context, here is how much work there was done during the same time period:

![image](https://github.com/near/nearcore/assets/6342444/12c72bc6-fda6-48a9-9dbd-aeda2872304e)

While the TX pool was not full, the network was able to saturate shard 3, as proven by >1s of compute per second in shard 3.
But then it hits this wall where only occasionally a tx would get in. (each tx is 300Tgas and about 600ms of compute cost, so even a small number of transactions produces non-trivial load)

How the load was produced is potentially relevant.
- I was using around 100 users but they were sharing 16 signer keys among each other. This can lead to nonce conflicts. Some of those conflicts are rejected at the RPC node. But due to the delay in execution, the RPC most of the time doesn't know that a nonce is invalid. In that case, users end up polling forever even though the transaction was already dropped form all validator tx pools.
- Users waited up to 30min on each transaction before retrying. Most transactions did hit that limit of 30mins. But if we could keep using the full capacity, all transactions should finish in less than 30min since the input rate was low enough that queue delay should be at most a few minutes. It could be due to the nonce conflicts, but I think there must be another factor since we no longer use full shard capacity.
- I have now started sending transactions (to the already full pools) that have no nonce conflicts. They still cannot get more shard capacity. (I'm getting about 1 transaction accepted per 8 blocks)

@akashin Can you think of potential deficiencies in how we remove transactions from the pool? I feel like we should be able to handle this situation better with the current implementation, so I kind of suspect there might be a bug / a case we forgot in how to remove transactions that we know will never be executed.

**Comment by jakmeier** [comment]
Oh another hint: I saw some `WARN network: Received invalid tx err=Expired` warnings on the validator nodes. But only one every few minutes, and sometimes there are no such warnings for several hours.

**Comment by akashin** [comment]
It would be interesting to see how many transactions are included at every chunk, how much gas we spend for converting receipts to transactions and how many delayed receipts we have. Spending gas there will likely decrease the compute usage temporarily, but then eventually we should be able to convert all transactions into delayed receipts and at that point get the high compute usage back.

**Comment by jakmeier** [comment]
Most chunks are completely empty, no transactions included at all.

![image](https://github.com/near/nearcore/assets/6342444/f8633b69-61b9-4175-ba73-be644fb59f31)

This was particularly weird with a full tx pool. But I have confirmed that I get this same behavior also with an empty tx pool.
So I am now thinking it is more likely to be an issue in how we send transactions or maybe just how we detect errors...

**Comment by jakmeier** [comment]
I am edging towards a point where I wonder if my time is spent well on trying to fix this.
Here is a brief summary of where I am with this, mostly to clear my mind by writing it down, but also as a potential checkpoint to interrupt / drop further work.


What exactly goes wrong?
- Somewhere, transactions are getting lost after 10-30s of the load starting.
- One user is always able to make progress, ALL other users are getting stuck no matter if there are 2 or 1000 users.
- From locust perspective:
  -  `broadcast_tx_commit` times out but doesn't give verification errors.
  - A follow-up call to `broadcast_tx_async` successfully returns a transaction ID.
  - Polling on that transaction ID never returns a result for the failing cases.
  - Successful cases work fine, which include the staring phase where many parallel transactions work fine.
- From the RPC nodes perspective:
  -  I see no messages about dropped transactions or anything else that looks relevant in the trace-level log
  - The number of forwarded transactions as recorded in the `near_transaction_received_non_validator` metric matches the submitted requests.
- From the validator node perspective:
  - I see nothing suspicious or useful in the log here, either. (we don't print received transaction IDs, so I can't check if specific transactions arrive without adding more logs)
  - It looks like the received transactions as recorded in `near_transaction_received_validator` is smaller than the number of forwarded messages. But it could also be an issue with metrics granularity playing tricks on me.
  - The number of network messages dropped is zero for all except duplicate chunk and block forwarding. ( `near_dropped_message_by_type_and_reason_count` is the metric I used here, with type!=Duplicate I get a flat 0 line in grafana)

What is confirmed to work:
- Sweat contract intialization with one oracle account per worker (already in master)
- Batches with ~100 account in a single receipt with a single action (already in master)
- Using different access keys per user of the per-worker oracle to avoid nonce conflicts (in my dev branch)
- Single user with large batches that combine multiple actions in a single receipt (in my dev branch)

For things to go wrong, the minimal requirements I reduced it to are:
- Batches need to be larger, which means they have to be split across multiple actions in a single receipt
  - I'm using 5x 60Tgas actions, each with an argument size of ~15kB, so the total tx size is still only around 75kB
  - For these large batches to be valid, I need to create and register about 1000 Sweat accounts first
- Multiple users must be submitting transactions at the same time
  - These can be sent from different oracle accounts as signer id, from different machines, to different RPC nodes

So the setup is still quite annoying for testing in quick iterations.

Why do we want to get this done?
- Right now, the main motivation is to finally have a end-to-end storage stress test that we can use to verify our ideas on how to optimize storage. Single receipts are already confirmed to hit the level of undercharging that the compute costs account for, which proves that the approach can work . But we need it in a sustained and stable way, otherwise it's not useful benchmark to run.

**Comment by jakmeier** [comment]
One more update: When I don't go through RPC nodes and instead send the transactions directly:
- Chunks are saturated for several minutes before most users hang again. (instead of 10-30s)
- 2 users (instead of 1) keep making progress

Still not sure what to make of it...

**Comment by jakmeier** [comment]
My latest theory is that it's due to skipped chunks and/or blocks. Perhaps, all transaction that were included in a chunk that the validator produced but was not included in the final chain go missing.
We do have code to remove and reintroduce transactions for such cases but I am not super familiar with that code. So maybe this is less robust than I assumed.

To test my hypothesis, I gave validators more time to apply chunks, hoping that this way nothing will be skipped even when we send those large transactions. I changes the following in config.json for all validators:
```yaml
    block_production_tracking_delay:  0.1s -> 0.5s
    min_block_production_delay:       0.6s -> 1.6s
    max_block_production_delay:       2s   -> 8s
    max_block_wait_delay:             6s   -> 16s
```
And it works! Mostly. I'm now getting a steady flow of saturated chunks. (see shard 3 gas usage, since the compute cost factor here is about 2.5, any chunks with more than 400Tgas are full.)

![image](https://github.com/near/nearcore/assets/6342444/6d842227-ba75-4f3c-bd21-0226982faaf0)

Unfortunately, there are still gaps of empty chunks in there. And I also still see some skipped chunks. So I still can't completely confirm (or deny) my hypothesis.

**Comment by bowenwang1996** [comment]
@jakmeier is it possible that when transaction pools are full, iterating through them becomes very slow, which may cause validators to miss chunks? Could we measure how much time is spend on `NightshadeRuntime::prepare_transactions`? cc @robin-near

**Comment by robin-near** [comment]
iirc chunk producers only get a chance to remove expired transactions when it's their turn to produce a chunk. So technically there can be a lot of garbage to get through, and each transaction would need to be verified (probably including the expired ones, filtered out by a nonce query?) which involves some sequential storage read operations. So technically I suppose it can be quite slow in theory.

**Comment by akashin** [comment]
> I'm using 5x 60Tgas actions, each with an argument size of ~15kB, so the total tx size is still only around 75kB

Actually, 75kB is quite close to 100kB that I've used in another problematic load test: https://github.com/near/nearcore-private/issues/81.
Do you happen to see any similar symptoms? I guess in your case the chain is still steadily growing, so not exactly the same behavior, but chunk skipping is definitely a suspect.

**Comment by jakmeier** [comment]
I'm back ti investigating this.

I tried to test if it is the same problem. I am running with verbose chain logs but I don't see any occurrences of `TooManyProcessingBlocks` in either of the two validators. So I don't think it is the same issue as described in https://github.com/near/nearcore-private/issues/81.

Will post more updates when I know more.

**Comment by jakmeier** [comment]
Looks like I finally pinned it down! Transactions are silently lost when their `block_hash` field references a block that hasn't been fully processed yet. Now I have consistently full blocks that replicate around 2x undercharging, sometimes even more.

The fix was easy, I just have to query the latest final block to set the block_hash field on the transaction. What we did so far is look at `/status` and read the `latest_hash` from there, which I'm not quite sure what the guarantees are their. Also no idea why that was chosen in our python cluster.py helper library, it uses the JSON RPC for most things, why go to the status page for this?

Anyway, let me try and explain what happens exactly.
I have to admit, I don't 100% understand how this problem happens, or why it consistently leads to exactly one user making progress while all other getting stuck. But I can clearly observe this in the logs:

On the chunk producer that loses the transaction:
- 16:47:48.943: the block hash `GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC` is in an incoming transaction at the chunk producer and it gets rejected as `Invalid tx: expired or from a different fork` ([code that rejects it because block hash is unknown](https://github.com/near/nearcore/blob/175e6ed030d0eba6ae1e9abfb749ae243e697968/chain/chain/src/store.rs#L523))
- 16:47:48.945 (2ms later) on the same node: `postprocess_block{height=132918472}: chain: Head updated to GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC at 132918472`
  - this confirms that when the tx should have been included, the header was still behind

Presumably the user got the block hash from the block producer itself that already knew about the new block ~250ms earlier.
Here some relevant logs of the block producer of `GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC`:
- 16:47:48.698 `handle_block_production:produce_block{next_height=132918472}:produce_block{next_height=132918472}: client: AccountId("v0.test.near") Producing block at height 132918472, parent 132918471 @ J3rUsQ, 4 new chunks`

- 16:47:48.701 `handle_block_production:produce_block{next_height=132918472}:start_process_block_impl{height=132918472}: chain: Some(AccountId("v0.test.near")) Process block GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC, is_caught_up: true`
- 16:47:48.702 `handle{handler="BlockResponse" actor="ClientActor" msg_type="BlockResponse"}:receive_block{me=Some(AccountId("v0.test.near")) prev_hash=J3rUsQD2jxi5v3tG2N7NkKioeMDGD5yut13edzSCT
11j hash=GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC height=132918472 peer_id=ed25519:BcjXwrPq9ScBN56cxBjYXhV62We6Cev9SV57cyMmtd42 was_requested=false}: chain: Process block header: GJJTWVYP8i9e7kWxPNttC7bLnJ5vtkNPhdterFCfG6NC at 132918472`

PRs and results will follow next week, now it's time for a weekend.
